Write to a single output file in parallel

t00sa opened the pull request:
This adds the ability to write to a single output file in parallel using MPI IO.
A file view is used to get each task to write to its own region of the file.

The following changes have been made to the existing classes to make
this change - and future file changes - easier to implement:

* a change to the Formatter signature to use `ostream` instances in
  place of `ofstream` instances.  This makes it possible to use a string
  to buffer the output instead of outputting it immediately
* a change to the Writer class to prevent the output stream from
  being exposed to the caller.  This allows the class to replace the
  internal IO method without changing the API, e.g. to use MPI IO
  instead of a C++ IO stream.

Testing and documentation have been updated to include the new
output mode.

andrewcoughtrie commented on documentation/Sphinx/user_guide/setting-up/setting-up.rst:
```diff
@@ -114,11 +114,23 @@ Environment Variables
+       
+     * **single**: All tasks write to a single global output file
+
+     If the environment variable is not set, **multi** mode is used.
```
We may want to change the default at some point as the single file output is likely to be the preferred method for most.

andrewcoughtrie replied:
Collated?

andrewcoughtrie commented on src/c++/mpi_context.cpp:
```diff
@@ -110,3 +113,56 @@ int meto::MPIContext::get_rank() { return comm_rank_; }
+  os << buffer.str();
+  os.flush();
+  os.close();
+
+#else // USE_VERNIER_MPI_STUB
+
+  int length;              // Local string length
+  int max_length;          // Global maximum string length
+  MPI_Datatype mpi_buffer; // Buffer as a contiguous item
+  MPI_File file_handle;    // MPI File accessor
+  MPI_Status status;       // Result of write
+
+  // Global maximum string size is required on every task to set up
+  // the custom data type
+  length = static_cast<int>(buffer.str().length());
+  MPI_Allreduce(&length, &max_length, 1, MPI_INT, MPI_MAX, comm_handle_);
+
+  MPI_Type_contiguous(max_length, MPI_CHAR, &mpi_buffer);
+  MPI_Type_commit(&mpi_buffer);
+
+  // Open the global output file and create a view for each task which
+  // represents a unique, non-overlapping region
+  MPI_File_open(comm_handle_, filename.c_str(),
+                MPI_MODE_CREATE | MPI_MODE_WRONLY, MPI_INFO_NULL, &file_handle);
+
+  MPI_File_set_view(file_handle, max_length * comm_rank_, MPI_CHAR, mpi_buffer,
+                    "native", MPI_INFO_NULL);
+
+  // Collective write operation
+  MPI_File_write_all(file_handle, buffer.str().c_str(), max_length, MPI_CHAR,
+                     &status);
+
+  // Tidy up resources
+  MPI_File_close(&file_handle);
+  MPI_Type_free(&mpi_buffer);
+
+#endif // USE_VERNIER_MPI_STUB
+
+  return;
+}
```
I think @mo-mglover ad my thoughts are both that this shouldn't be kept in the `mpi_context.cpp` file it would probably be preferable to wrap/dummy the actual MPI calls and have the actually use in the `single.cpp` file. This would keep the MPI usage contained away from our usage in the writers.

t00sa replied:
I've created an issue to remind us to come back to this

andrewcoughtrie reviewed (changes requested):
The main concept looks really good, I think we need to do a bit of refactoring of where the code lives but the overall mechanism looks great. I also think we could do with a small system test that checks the the file being output is correct, to begin with that could just check the filename is correct, see [test-tags.cpp](https://github.com/MetOffice/Vernier/blob/main/tests/system_tests/c%2B%2B/test-tags.cpp).

Otherwise great addition.

t00sa commented:
Things still to do:

- [ ] Change the name of the output file from `global` to `collated`
- [ ] Add a system test to confirm that an output file is being generated in the right place
- [ ] Add a separator/identifier, e.g. `rank X of Y` to identify each section of output

t00sa commented:
Right, all changes made.  I think it's good to go for another review.

andrewcoughtrie commented on src/c++/formatter.cpp:
```diff
@@ -45,7 +45,12 @@ meto::Formatter::Formatter() {
+void meto::Formatter::execute_format(std::ostream &os, hashvec_t hashvec,
+                                     MPIContext &mpi_context) {
+  // Add an MPI task identifier to each output file
+  os << "\n"
+     << "MPI task " << mpi_context.get_rank() << " of "
+     << mpi_context.get_size() << "\n";
```
The output here comes out as `MPI task 0 of 1` I'm wondering whether it is better to have `MPI task 1 of 1 : rank ID 0` this would give both the integer value of the total whilst also having the rank id without the confusion of the total number being offset by one.

andrewcoughtrie replied:
That's a simpler alternative (or at least simpler to read if not to implement 😁)

andrewcoughtrie replied:
That would work well, it would make it clearer in the multi file output as well as each file would still tell you how many ranks there are in total.

andrewcoughtrie commented on tests/system_tests/c++/test-output-files.cpp:
```diff
@@ -0,0 +1,121 @@
+    path = "vernier-output-collated";
+  } else {
+    std::cerr << "unknown mode" << std::endl;
+    return false;
```
I really don't like multiple return statements in a function, I think we'd be better off using a `GOTO` statement :stuck_out_tongue_winking_eye:

andrewcoughtrie replied:
It's a test, we have no logger to error out so return is fine.

t00sa commented:
I haven't changed the output reporting or the multiple returns, but I've changed formatting and added a multiple MPI task test.

t00sa commented:
I think the failures were probably down to the MPI tasks racing with each other, so I've added code to limit the check to rank zero and renamed the output files for good measure.
